File: src/main/java/edu/kruchon/natural/lang/parser/syntax/TestParagraphSyntaxTreeParser.java
package edu.kruchon.natural.lang.parser.syntax;

import edu.stanford.nlp.ie.util.RelationTriple;
import edu.stanford.nlp.ling.CoreAnnotations;
import edu.stanford.nlp.ling.IndexedWord;
import edu.stanford.nlp.naturalli.NaturalLogicAnnotations;
import edu.stanford.nlp.pipeline.Annotation;
import edu.stanford.nlp.pipeline.StanfordCoreNLP;
import edu.stanford.nlp.semgraph.SemanticGraph;
import edu.stanford.nlp.trees.GrammaticalRelation;

import java.util.Collection;
import java.util.Comparator;
import java.util.List;
import java.util.Properties;
import java.util.Set;
import java.util.stream.Collectors;

public enum TestParagraphSyntaxTreeParser {
    INSTANCE;

    private final StanfordCoreNLP pipeline;

    TestParagraphSyntaxTreeParser() {
        Properties props = new Properties();
        props.setProperty("annotators", "tokenize,ssplit,pos,lemma,depparse,natlog,openie");
        pipeline = new StanfordCoreNLP(props);
    }

    public Triplet parse(String paragraph) {
        Annotation doc = new Annotation(paragraph);
        pipeline.annotate(doc);
        RelationTriple relationTriple = doc.get(CoreAnnotations.SentencesAnnotation.class).stream()
                .map(
                        coreMap -> coreMap.get(NaturalLogicAnnotations.RelationTriplesAnnotation.class)
                )
                .flatMap(
                        Collection::stream
                )
                .max(
                        Comparator.comparingDouble(f -> f.confidence)
                ).orElseThrow(
                        () -> new RuntimeException("Relation triple was not found")
                );
        SemanticGraph dependencyTree = relationTriple.asDependencyTree().orElseThrow(() -> new RuntimeException("Error building dependency tree"));
        IndexedWord relationshipNode = dependencyTree.getFirstRoot();
        IndexedWord object = dependencyTree.getChildrenWithReln(relationshipNode, GrammaticalRelation.valueOf("obj")).iterator().next();
        return new Triplet(relationTriple.subjectLemmaGloss(), relationTriple.relationLemmaGloss(), processNode(object, dependencyTree));
    }

    private Parameter processNode(IndexedWord node, SemanticGraph dependencyTree) {
        // todo fix the case with needless nmods
        Set<IndexedWord> childrenNodes = dependencyTree.getChildren(node);
        List<Parameter> childrenParameterNodes = childrenNodes.stream()
                .filter(childrenNode -> !dependencyTree.getChildren(childrenNode).isEmpty())
                .sorted(Comparator.comparingInt(IndexedWord::beginPosition))
                .map(childrenNode -> processNode(childrenNode, dependencyTree))
                .collect(Collectors.toList());
        List<String> childrenValueNodes = childrenNodes.stream()
                .filter(childrenNode -> dependencyTree.getChildren(childrenNode).isEmpty())
                .sorted(Comparator.comparingInt(IndexedWord::beginPosition))
                .map(IndexedWord::originalText)
                .collect(Collectors.toList());
        return new Parameter(node.originalText(), childrenValueNodes, childrenParameterNodes);
    }
}

File: src/main/kotlin/edu/kruchon/natural/lang/parser/syntax/TestParagraphSyntaxTreeParser.kt
package edu.kruchon.natural.lang.parser.syntax

import edu.stanford.nlp.ling.CoreAnnotations.SentencesAnnotation
import edu.stanford.nlp.ling.IndexedWord
import edu.stanford.nlp.naturalli.NaturalLogicAnnotations.RelationTriplesAnnotation
import edu.stanford.nlp.pipeline.Annotation
import edu.stanford.nlp.pipeline.StanfordCoreNLP
import edu.stanford.nlp.semgraph.SemanticGraph
import edu.stanford.nlp.trees.GrammaticalRelation
import java.util.*

object TestParagraphSyntaxTreeParser {

    private val pipeline: StanfordCoreNLP

    init {
        val props = Properties()
        props.setProperty("annotators", "tokenize,ssplit,pos,lemma,depparse,natlog,openie")
        pipeline = StanfordCoreNLP(props)
    }

    fun parse(paragraph: String): Triplet {
        val doc = Annotation(paragraph)
        pipeline.annotate(doc)
        val relationTriple = checkNotNull(doc.get(SentencesAnnotation::class.java)
                .map { it.get(RelationTriplesAnnotation::class.java) }
                .flatten()
                .maxByOrNull { it.confidence })
        val dependencyTree: SemanticGraph = relationTriple.asDependencyTree().orElseThrow { RuntimeException("Error building dependency tree") }
        val relationshipNode = dependencyTree.firstRoot
        val `object`: IndexedWord = dependencyTree.getChildrenWithReln(relationshipNode, GrammaticalRelation.valueOf("obj")).iterator().next()
        return Triplet(relationTriple.subjectLemmaGloss(), relationTriple.relationLemmaGloss(), processNode(`object`, dependencyTree))
    }

    private fun processNode(node: IndexedWord, dependencyTree: SemanticGraph): Parameter {
        // todo fix the case with needless nmods
        val childrenNodes = dependencyTree.getChildren(node)
        val childrenParameterNodes = childrenNodes
                .filter { childrenNode -> dependencyTree.getChildren(childrenNode).isNotEmpty() }
                .sortedBy {
                    it.beginPosition()
                }
                .map { childrenNode -> processNode(childrenNode, dependencyTree) }
        val childrenValueNodes = childrenNodes
                .filter { childrenNode -> dependencyTree.getChildren(childrenNode).isEmpty() }
                .sortedBy {
                    it.beginPosition()
                }
                .map(IndexedWord::originalText)
        return Parameter(node.originalText(), childrenValueNodes, childrenParameterNodes)
    }
}